### PR TITLE
nixos/nixpkgs/lib/strings: add stringToValidNixPath

### PR DESCRIPTION
**Pull request by rpearce**
(This is my first time contributing like this to nixpkgs... please be gentle)

###### Motivation for this change
When using tools like `nixpkgs.lib.attrsets.mapAttrsToList`
and trying to build derivations where nix store file names
are to be created, set attributes like `".stack/config.yml"`
are invalid, so `sanitizeNixPath` cleans them up to allow
them to be created in the nix store.

I did not include `-` as a valid character because of [rycee's reasoning here](https://github.com/rycee/home-manager/blob/b9d4f55228befb6ea5b1888a0e472b40776b3cc8/modules/lib/strings.nix#L9-L11).

If no prefix is provided, the string `safePrefix_` is used.

###### Things done
* adds `strings.sanitizeNixPath`
* adds tests for `strings.sanitizeNixPath`

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

* * *

I'm not sure how to run the tests locally and could use some help doing that.

### CODE REVIEW

File: lib/strings.nix [hunk 1]
```diff
@@ -678,4 +678,31 @@ rec {
...
+     Type: stringToValidNixPath :: string -> string -> string
+
+     Example:
+       stringToValidNixPath "foo_" ".stack/config.yml"
+       => "foo_.stackconfig.yml"
+       stringToValidNixPath "  " ".stack/config.yml"
+       => "safePrefix_.stackconfig.yml"
+  */
+
+  stringToValidNixPath = prefix: path:
```
Comment (Mic92) [inline lib/strings.nix]: Does this sounds better?

```suggestion
  sanitizeNixPath = prefix: path:
```

Comment (rpearce) [reply lib/strings.nix]: That sounds much better!

**Comment by Infinisil** [comment]
I also have an implementation of this in https://github.com/NixOS/nixpkgs/pull/78640, see https://github.com/NixOS/nixpkgs/pull/78640/commits/30ad4eea5bff8a42c102a5483ddd055fd3c0799a

I could put that into its own PR

**Comment by rpearce** [comment]
@Infinisil If that's better than this, then by all means go for it. Or if you think I could adjust this to incorporate some of your concerns, like length, then I'm happy to update this and accomplish the same goal

File: lib/strings.nix [hunk 2]
```diff
@@ -678,4 +678,31 @@ rec {
...
+     Example:
+       sanitizeNixPath "foo_" ".stack/config.yml"
+       => "foo_.stackconfig.yml"
+       sanitizeNixPath "  " ".stack/config.yml"
+       => "safePrefix_.stackconfig.yml"
+  */
+
+  sanitizeNixPath = prefix: path:
+    let
+      safePrefix = if replaceStrings [" "] [""] prefix == "" then "safePrefix_" else prefix;
```
Comment (rpearce) [inline lib/strings.nix]: Should prefixing be left to consumers to figure out themselves?

**Comment by Infinisil** [comment]
I'm obviously biased towards my solution :). Some arguments for my solution are that it uses regex-based replacement, which should be faster than `builtins.replaceStrings` and the code is very clear as to what it does.

**Comment by rpearce** [comment]
@Infinisil If you get that into a PR, I'll test it out for you with what I'm testing with on my side, and if we get that to :+1:, I'll close this. How's that sound?

**Comment by Infinisil** [comment]
Done that :) https://github.com/NixOS/nixpkgs/pull/83241

**Pull request closed without merging**